src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,401 @@
+         * <p>
+         * This supports the use-case where a label can be applied to the IKM but the actual value
+         * of the IKM is not yet available.
+         *
+         * @param ikm
+         *     the ikm value (null or empty values will not be added)
+         *
+         * @return a new {@code Extract} object
+         */
+        public Builder addIKM(byte[] ikm) {
+            if (ikm != null && ikm.length != 0) {
+                return addIKM(new SecretKeySpec(ikm, "RAW"));
+            } else {
+                return this;
+            }
+        }
+
+        /**
+         * {@code addSalt} may be called when the salt value is to be assembled piece-meal or if
+         * part of the salt is to be supplied by a hardware crypto device. This method appends to
+         * the existing list of values or creates a new list if there are none yet.
+         * <p>
+         * This supports the use-case where a label can be applied to the salt but the actual value
+         * of the salt is not yet available.
+         *
+         * @param salt
+         *     the salt value (null values will not be added)
+         *
+         * @return a new {@code Extract} object
+         */
+        public Builder addSalt(SecretKey salt) {
+            if (salt != null) {
+                salts.add(salt);
+            }
+            return this;
+        }
+
+        /**
+         * {@code addSalt} may be called when the salt value is to be assembled piece-meal or if
+         * part of the salt is to be supplied by a hardware crypto device. This method appends to
+         * the existing list of values or creates a new list if there are none yet.
+         * <p>
+         * This supports the use-case where a label can be applied to the salt but the actual value
+         * of the salt is not yet available.
+         *
+         * @param salt
+         *     the salt value (null or empty values will not be added)
+         *
+         * @return a new {@code Extract} object
+         */
+        public Builder addSalt(byte[] salt) {
+            if (salt != null && salt.length != 0) {
+                return addSalt(new SecretKeySpec(salt, "RAW"));
+            } else {
+                return this;
+            }
+        }
+    }
+
+    /**
+     * Static helper-method that may be used to initialize a {@code Builder} with an empty
+     * {@code Extract}
+     * <p>
+     * Note: one or more of the methods {@code addIKM} or {@code addSalt} should be called next,
+     * before calling build methods, such as {@code Builder.extractOnly()}
+     *
+     * @return a {@code Builder} to mutate
+     */
+    static Builder extract() {
+        return new Builder().createExtract();
+    }
+
+    /**
+     * Static helper-method that may be used to initialize an {@code Expand} object
+     *
+     * @param prk
+     *     the PRK (may be null)
+     * @param info
+     *     the info (may be null)
```

Comment:
I know you use a null `prk` in `ExtractExpand`, but this method is public available for the Expand-Only mode and we don't want end users to provide a null here.

For `info`, I'd rather allow empty input and reject null.